teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1089 @@
+
+For most other examples, the `StepAction` itself will not require a `SecurityContext`. However, the pipeline authors may require setting `SecurityContext` (e.g. running as non-root) for all their underlying `Tasks`. The control of `SecurityContext` therefore needs to be available to both `StepActions` and `Tasks`.  
+
+However, this requires a hierarchy in cases where both `StepActions` and `Tasks` set this value to something. Since we expect the authors to set this value only for `StepActions` that actually need it because the underlying image dictates it, we take the precedence of the `StepAction` over what is set in the `Task` during resolution. Authors certainly have the ability to set this to arbitrary things even when it is not required but that would make the `StepAction` less reusable.
+
+#### Validating StepActions and Steps
+When `StepActions` are applied directlty on the cluster, we will invoke a webhook validation of the API. This would be similar to validating inlined `Steps` with the addition of `Params` and `Results`. The same validation will also be envoked if the `StepAction` is fetched via the remote resolvers by triggering a `dry-run` submission.
+
+If the `Step` references `StepActions`, it cannot also use the same fields that fall under `StepActions` (i.e. `image`, `command`, `args`, `env`, `script`, `volumeMounts`). This would lead to validation error. If the `Step` is not using `StepActions`, it cannot declare `params` and `results` since it is being inlined. This will also result in a validation error. The Step should only provide `params` ad `results` if referencing a `StepAction`.
+
+After the `StepAction's` spec is fetched, it will efectively convert the `Step` referencing a `StepAction` to an inlined `Step` in the status of the `TaskRun`. We do a similar thing when we resolve `PipelinRef` and `TaskRef`. After this point, the remainder of the validation happens as if it was an inlined `Step` (i.e. no additional validation is required). 
+
+#### Managing API and feature versions
+The `StepAction` CRD will be introduced under `v1alpha1` API version and promoted to `v1beta1` and finally to `v1` following the [API compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md).
+
+At the same time, we also need to introduce a feature flag for enabling referencing of `StepActions` from `Steps`. [TEP-0138](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md) introduced per-feature-flags. Following that, we will introduce a feature flag called `enable-step-actions`, dedicated for this feature. 
+
+Since both the CRD and the feature flag, are being introduced at the same time, we propose syncing the advancement of the feature flag `enable-step-actions` and the API stability level for `StepActions`.
```

Comment:
Thanks @chitrangpatel. As I understand, this means that:
- when `enable-step-actions` is alpha, the `v1` Task will be able to use `v1alpha1.StepActions`
- when `enable-step-actions` is beta, the `v1` Task will be able to use `v1beta1.StepActions`. Will support for the `v1alpha1.StepActions` cease immediately? Will we provide a conversion webhook/logic?
- when `enable-step-actions` is stable, the `v1` Task will be able to use `v1.StepActions`. How will we handle support for `v1beta1.StepAction`?

Reply:
Thanks for the detailed questions. It helped me dig into more details. I updated it. PTAL. I think we will have a deprecation window between each promotion so that users can make the transition from `v1alpha1 --> v1beta1 --> v1`. And in the interim, we will have web hook conversion to convert between the two so that the TaskRun controller can work with only one of the two types.